# 1. Reverse only EVEN digits

In [1]:
num = int(input("Enter number: "))

digits = list(str(num))   # convert to list of characters

# extract even digits
even_digits = [d for d in digits if int(d) % 2 == 0]

even_digits.reverse()    # reverse only even digits

index = 0

for i in range(len(digits)):
    if int(digits[i]) % 2 == 0:
        digits[i] = even_digits[index]
        index += 1

print("Result:", "".join(digits))

Enter number: 123456
Result: 163452


# 2. Reverse number without using extra variable

In [2]:
num = int(input("Enter number: "))
rev = 0

while num:
    rev = rev * 10 + num % 10
    num //= 10

print(rev)

Enter number: 123
321


# 3. Reverse words but keep order of words same

In [3]:
# Example: "I love Python" → "I evol nohtyP"
s = input("Enter sentence: ")

result = " ".join(word[::-1] for word in s.split())

print(result)

Enter sentence: I love Python
I evol nohtyP


# 4. Reverse sentence but keep characters same

Example: "I love Python" → "Python love I"

In [4]:
s = input("Enter sentence: ")

print(" ".join(s.split()[::-1]))

Enter sentence: I love Python
Python love I


# 5. Reverse only vowels in a string

Example: "hello" → "holle"

In [5]:
s = list(input("Enter string: "))

vowels = "aeiouAEIOU"

# extract vowels
v = [ch for ch in s if ch in vowels]
v.reverse()

index = 0

for i in range(len(s)):
    if s[i] in vowels:
        s[i] = v[index]
        index += 1

print("".join(s))

Enter string: hello
holle


# 6. Reverse digits of number but keep sign same

Example: -123 → -321

In [6]:
num = int(input("Enter number: "))

sign = -1 if num < 0 else 1
num = abs(num)

rev = 0

while num:
    rev = rev * 10 + num % 10
    num //= 10

print(sign * rev)

Enter number: -123
-321


# 7. Reverse list in-place (no extra space)

In [7]:
lst = [1, 2, 3, 4, 5]

left = 0
right = len(lst) - 1

while left < right:
    lst[left], lst[right] = lst[right], lst[left]
    left += 1
    right -= 1

print(lst)

[5, 4, 3, 2, 1]


# 8. Reverse only odd index elements

Example: [1,2,3,4,5] → [1,4,3,2,5]

In [8]:
lst = [1, 2, 3, 4, 5]

odd_vals = [lst[i] for i in range(len(lst)) if i % 2 != 0]
odd_vals.reverse()

index = 0

for i in range(len(lst)):
    if i % 2 != 0:
        lst[i] = odd_vals[index]
        index += 1

print(lst)

[1, 4, 3, 2, 5]


# 9. Reverse until a specific position

Example: 123456, k=3 → 321456

In [9]:
num = input("Enter number: ")
k = int(input("Enter k: "))

result = num[:k][::-1] + num[k:]

print(result)

Enter number: 123456
Enter k: 3
321456


# 10. Reverse number and remove trailing zeros

Example: 1200 → 21

In [10]:
num = int(input("Enter number: "))

rev = 0

while num:
    rev = rev * 10 + num % 10
    num //= 10

print(rev)
# Automatically removes leading zeros after reverse

Enter number: 1200
21
